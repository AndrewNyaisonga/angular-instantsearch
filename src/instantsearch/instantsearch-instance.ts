import { Injectable } from "@angular/core";
import instantsearch from "instantsearch.js/es";

@Injectable()
export class NgISInstance {
  public appId: string;
  public apiKey: string;
  public indexName: string;

  public numberLocale?: string;
  public searchFunction?: () => void;
  public createAlgoliaClient?: () => object;
  public searchParameters?: object | void;
  public urlSync?:
    | boolean
    | {
        mapping?: object;
        threshold?: number;
        trackedParameters?: string[];
        useHash?: boolean;
        getHistoryState?: () => object;
      };

  private instance?: {
    start: () => void;
  };

  public init(config: object) {
    this.instance = instantsearch(config);
  }

  public start() {
    this.instance.start();
  }
}
