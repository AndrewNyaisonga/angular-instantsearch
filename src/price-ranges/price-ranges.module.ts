import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisFooterModule } from "../footer/footer.module";
import { NgAisHeaderModule } from "../header/header.module";
import { NgAisPriceRanges } from "./price-ranges";

@NgModule({
  declarations: [NgAisPriceRanges],
  entryComponents: [NgAisPriceRanges],
  exports: [NgAisPriceRanges],
  imports: [CommonModule, NgAisHeaderModule, NgAisFooterModule]
})
export class NgAisPriceRangesModule {}
