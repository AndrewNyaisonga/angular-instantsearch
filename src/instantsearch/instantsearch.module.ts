import { CommonModule } from "@angular/common";
import { ModuleWithProviders, NgModule } from "@angular/core";

import { NgISInstantSearch } from "./instantsearch";
import { NgISInstance } from "./instantsearch-instance";

@NgModule({
  declarations: [NgISInstantSearch],
  entryComponents: [NgISInstantSearch],
  exports: [NgISInstantSearch],
  imports: [CommonModule]
})
export class NgISInstantSearchModule {
  public static forRoot(): ModuleWithProviders {
    return {
      ngModule: NgISInstantSearchModule,
      providers: [NgISInstance]
    };
  }
}
