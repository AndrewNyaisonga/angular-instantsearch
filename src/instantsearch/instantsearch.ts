import { AfterViewInit, Component, Input, OnInit } from "@angular/core";

import { NgISInstance } from "./instantsearch-instance";

@Component({
  selector: "ngis-instantsearch",
  template: `<ng-content></ng-content>`
})
export class NgISInstantSearch implements AfterViewInit, OnInit {
  @Input() public config: InstantSearchConfig;

  constructor(private searchInstance: NgISInstance) {}

  public ngOnInit() {
    this.searchInstance.init(this.config);
  }

  public ngAfterViewInit() {
    this.searchInstance.start();
  }
}
