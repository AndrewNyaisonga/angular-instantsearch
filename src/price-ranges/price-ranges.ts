import { Component, Input } from "@angular/core";
import { connectPriceRanges } from "instantsearch.js/es/connectors";
import { noop } from "lodash";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";
import { bem } from "../utils";

const cx = bem("PriceRanges");

@Component({
  selector: "ng-ais-price-ranges",
  template: `
    <div class="${cx()}">
      <ng-ais-header [header]="header" className="${cx(
        "header"
      )}"></ng-ais-header>

      <div class="${cx("body")}">
        <ul class="${cx("list")}">
          <li
            [ngClass]="{
              '${cx("item")}': true,
              '${cx("item", "selected")}': item.isRefined
            }"
            *ngFor="let item of state.items"
            (click)="handleClick($event, item)"
          >
            {{formatLabel(item)}}
          </li>
        </ul>
      </div>

      <ng-ais-footer [footer]="footer" className="${cx(
        "footer"
      )}"></ng-ais-footer>
    </div>
  `
})
export class NgAisPriceRanges extends BaseWidget {
  // render options
  @Input() public currency: string = "$";
  @Input() public formatLabel = this.defaultFormatLabel;

  // connector options
  @Input() public attributeName: string;

  public state = {
    items: [],
    refine: noop
  };

  constructor(searchInstance: NgAisInstance) {
    super(searchInstance);
  }

  public ngOnInit() {
    this.createWidget(connectPriceRanges, {
      attributeName: this.attributeName
    });
    super.ngOnInit();
  }

  public defaultFormatLabel({ from, to }) {
    if (to === undefined) {
      return `>= ${this.currency} ${from}`;
    }
    if (from === undefined) {
      return `<= ${this.currency} ${to}`;
    }
    return `${this.currency} ${from} - ${this.currency} ${to}`;
  }

  public handleClick(event, item) {
    event.preventDefault();
    this.state.refine(item);
  }
}
