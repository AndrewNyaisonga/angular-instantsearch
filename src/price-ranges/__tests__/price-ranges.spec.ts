import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../../instantsearch/instantsearch.module";
import { NgAisPriceRangesModule } from "../price-ranges.module";
import { NgAisPriceRanges } from "../price-ranges";

jest.mock("../../base-widget");

const defaultState = {
  items: [{ from: 1 }, { from: 10, to: 20 }, { to: 50 }],
  refine: jest.fn()
};

const render = (state?: {}) => {
  const fixture = TestBed.createComponent(NgAisPriceRanges);

  if (state) {
    fixture.componentInstance.updateState({ ...defaultState, ...state }, false);
  }

  fixture.detectChanges();
  return fixture;
};

describe("PriceRanges", () => {
  beforeEach(() =>
    TestBed.configureTestingModule({
      declarations: [],
      imports: [NgAisInstantSearchModule.forRoot(), NgAisPriceRangesModule]
    })
  );

  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("renders refined values", () => {
    const fixture = render({ items: [{ to: 20, isRefined: true }] });
    expect(fixture).toMatchSnapshot();
  });

  it("should call `refine()` on item click", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const [, secondItem] = fixture.debugElement.nativeElement.querySelectorAll(
      "li"
    );
    secondItem.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.items[1]);
  });
});
