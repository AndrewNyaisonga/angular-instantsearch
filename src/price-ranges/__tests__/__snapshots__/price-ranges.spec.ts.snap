// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceRanges renders markup with state 1`] = `
<ng-ais-price-ranges
  currency={[Function String]}
  cx={[Function Function]}
  formatLabel={[Function Function]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-PriceRanges"
  >
    
        
    <ng-ais-header
      ng-reflect-class-name="ais-PriceRanges-header ais-hea"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-PriceRanges-body ais-body"
    >
      
          
      <ul
        class="ais-PriceRanges-list"
      >
        
            
        
        <li
          class="ais-PriceRanges-item"
        >
          
              &gt;= $ 1
            
        </li>
        <li
          class="ais-PriceRanges-item"
        >
          
              $ 10 - $ 20
            
        </li>
        <li
          class="ais-PriceRanges-item"
        >
          
              &lt;= $ 50
            
        </li>
        
          
      </ul>
      
        
    </div>
    
  
        
    <ng-ais-footer
      ng-reflect-class-name="ais-PriceRanges-footer ais-foo"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-price-ranges>
`;

exports[`PriceRanges renders markup without state 1`] = `
<ng-ais-price-ranges
  currency={[Function String]}
  cx={[Function Function]}
  formatLabel={[Function Function]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-PriceRanges"
  >
    
        
    <ng-ais-header
      ng-reflect-class-name="ais-PriceRanges-header ais-hea"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-PriceRanges-body ais-body"
    >
      
          
      <ul
        class="ais-PriceRanges-list"
      >
        
            
        
        
          
      </ul>
      
        
    </div>
    
  
        
    <ng-ais-footer
      ng-reflect-class-name="ais-PriceRanges-footer ais-foo"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-price-ranges>
`;

exports[`PriceRanges renders refined values 1`] = `
<ng-ais-price-ranges
  currency={[Function String]}
  cx={[Function Function]}
  formatLabel={[Function Function]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-PriceRanges"
  >
    
        
    <ng-ais-header
      ng-reflect-class-name="ais-PriceRanges-header ais-hea"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-PriceRanges-body ais-body"
    >
      
          
      <ul
        class="ais-PriceRanges-list"
      >
        
            
        
        <li
          class="ais-PriceRanges-item ais-PriceRanges-item--selected"
        >
          
              &lt;= $ 20
            
        </li>
        
          
      </ul>
      
        
    </div>
    
  
        
    <ng-ais-footer
      ng-reflect-class-name="ais-PriceRanges-footer ais-foo"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-price-ranges>
`;
